lib/Dialect/StableHLO/Transforms/FlattenComposite.cpp

```diff
@@ -0,0 +1,186 @@
+// SPDX-FileCopyrightText: (c) 2025 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "stablehlo/dialect/StablehloOps.h"
+#include "ttmlir/Dialect/StableHLO/Transforms/Passes.h"
+#include "ttmlir/Dialect/StableHLO/Utils/ShardingUtils.h"
+#include "ttmlir/Dialect/StableHLO/Utils/ShardyUtils.h"
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/IRMapping.h"
+#include "llvm/Support/Error.h"
+
+namespace mlir::tt::stablehlo {
+#define GEN_PASS_DEF_FLATTENCOMPOSITEPASS
+#include "ttmlir/Dialect/StableHLO/Transforms/Passes.h.inc"
+
+// Inline a single stablehlo.composite op. Returns success if it was flattened.
+static mlir::LogicalResult
+flattenOneComposite(mlir::stablehlo::CompositeOp comp,
+                    mlir::SymbolTable &symTable, mlir::OpBuilder &builder) {
+  mlir::Operation *op = comp.getOperation();
+  mlir::MLIRContext *context = builder.getContext();
+
+  // 0) Read metadata from the original composite.
+  mlir::StringAttr origName = mlir::StringAttr::get(context, comp.getName());
+  // Preserve original composite attributes.
+  mlir::DictionaryAttr origCompAttrs = comp.getCompositeAttributes();
+  if (!origCompAttrs) {
+    // Create an empty attribute if none exists.
+    origCompAttrs = mlir::DictionaryAttr::get(context);
+  }
+
+  // 1) Resolve callee from 'decomposition' (SymbolRefAttr).
+  // stablehlo.composite "name" %arg {decomposition = @foo}
+  auto decompAttr = op->getAttrOfType<mlir::SymbolRefAttr>(
+      sharding_utils::kDecompositionAttr);
+  if (!decompAttr) {
+    comp.emitOpError()
+        << "missing required SymbolRefAttr attribute '"
+        << sharding_utils::kDecompositionAttr
+        << "' (stablehlo.composite requires a 'decomposition' target).";
+    return mlir::failure();
+  }
+
+  // SymbolRefAttr may be nested; use the leaf for func name.
+  mlir::StringAttr leaf = decompAttr.getLeafReference();
+  mlir::func::FuncOp callee = symTable.lookup<mlir::func::FuncOp>(leaf);
+
+  if (!callee) {
+    comp.emitOpError() << "failed to resolve callee function '" << leaf
+                       << "' referenced by attribute '"
+                       << sharding_utils::kDecompositionAttr << "' (full ref: '"
+                       << decompAttr
+                       << "'). Please ensure the symbol is defined and visible "
+                          "to the symbol table.";
+    return mlir::failure();
+  } else if (callee.getBody().empty()) {
```

Comment:
:warning: **llvm\-else\-after\-return** :warning:
do not use `` else `` after `` return ``
```suggestion
  } if (callee.getBody().empty()) {
```